replit/upm
# Datadog trace with distributed tracing

# Why?

Want to measure/trace performance of upm code paths, especially upm lock and upm guess to find bottlenecks.

## Changes

1. Implemented code to take existing DD trace ID and span ID passed from parent process via `DD_TRACE_ID` and `DD_SPAN_ID` variables and inject them into the span context.
2. Add context parameter to functions we want traced.
3. Added span tracing code to various places.

## Testing

Instructions for manual local testing:

1. [Install datadog agent locally](https://docs.datadoghq.com/agent/) (you'll need an API key, you can create a new one or ask me privately for mine)
2. Make sure the agent is running via something like [sudo datadog-agent status](https://docs.datadoghq.com/agent/basic_agent_usage/ubuntu/?tab=agentv6v7)
4. Compile upm and make sure it's in `PATH`. I do this with `nix develop; make install`
5. Make a test directory with a `index.js` containing a require statement, say: `const ai = require('@replit/ai');`
6. cd into that test directory
7. exec the `./internal/trace/test_cli/test_cli.go` file with something like `go run ../internal/trace/test_cli/test_cli.go`
8. In the datadog APM trace UI, look for `service:upm_wrapper` and you should see something like:
![Screenshot from 2023-11-06 12-17-05](https://github.com/replit/upm/assets/54303/d66ebeb3-4807-4d9b-bb4b-9fdfa0b71ebd)
